posts/inside-rust/2024-10-01-this-development-cycle-in-cargo-1.82.md

```diff
@@ -0,0 +1,530 @@
+```toml
+[resolver]
+something-like-precedence = "something-like-rust-version"
+```
+to
+```toml
+[resolver]
+incompatible-rust-version = "fallback"
+```
+([#14296](https://github.com/rust-lang/cargo/pull/14296))
+
+With that in place and not aware of any other blockers,
+we put out a new round of [call-for-testing](https://github.com/rust-lang/cargo/issues/13873#issuecomment-2276647842).
+We're grateful for the feedback we've received!
+
+A particularly confusing aspect of the `fallback` strategy is that Cargo may select a dependency that is either
+- Too new for your `package.rust-version` because no MSRV-compatible version was available and so we did a `fallback` to an MSRV-incompatible one
+- Older than is needed by your package if your version requirement is too liberal and your package doesn't set `package.rust-version` but something else in your workspace does
+
+After several tweaks to Cargo's locking output (
+[#14401](https://github.com/rust-lang/cargo/pull/14401)
+[#14440](https://github.com/rust-lang/cargo/pull/14440).
+[#14445](https://github.com/rust-lang/cargo/pull/14445).
+[#14457](https://github.com/rust-lang/cargo/pull/14440).
+[#14459](https://github.com/rust-lang/cargo/pull/14459).
+[#14461](https://github.com/rust-lang/cargo/pull/14461).
+[#14471](https://github.com/rust-lang/cargo/pull/14471).
+), we now report to users
+- when Cargo selected a dependency version that is incompatible with the MSRV of workspace members that transitively depend on it
+- when a newer version of a package is available that is compatible with the MSRV of workspace members that transitively depend on it
+
+![rendering of cargo update](../../../../images/inside-rust/2024-10-01-this-development-cycle-in-cargo-1.82/update.stderr.term.svg)
+
+<!-- team meeting: 2024-08-20 -->
+<!-- team meeting: 2024-08-27 -->
+The testing feedback was more than `fallback` behavior being confusing but that it can fail to provide the benefits of an MSRV-aware resolver when in a workspace with multiple MSRVs ([#14414](https://github.com/rust-lang/cargo/issues/14414)).
+The ideal state would be that when resolving a dependency, we only consider the MSRVs of the workspace members that transitively depend on it.
+The problem is that we don't know every path to a dependency until the resolve is done.
+We select a version of a dependency when we first come across it and then only try another version if the selected version is rejected by another path to it.
+If we `deny`d dependency versions with an incompatible MSRV, with enough work we could hit the ideal case.
+The problem is how much work this would take and that there are times packages intentionally select MSRV-incompatible dependencies (e.g. features with a different MSRV, not caring about MSRV for dev-dependencies, etc).
+Our `fallback` strategy requires that we pick a "good enough" version when we version come across a dependency because another path to it cannot reject it and cause a new version to be selected.
```

Comment:
```suggestion
Our `fallback` strategy requires that we pick a "good enough" version when we come across a dependency version because another path to it cannot reject it and cause a new version to be selected.
```